EIPS/eip-epbs.md

```diff
@@ -0,0 +1,260 @@
+- `process_deposit_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- `process_withdrawal_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- `process_consolidation_request` is removed from `process_operations` and deferred until `process_execution_payload`. 
+- A new `process_payload_attestation` is added to `process_operations`, this function validates the payload timeliness attestations broadcast by the PTC members. 
+- `process_execution_payload` is now called as a separate helper when receiving a `SignedExecutionPayloadEnvelope` on the P2P layer. This function in particular checks that the HTR of the resulting beacon state coincides with the committed one in the payload envelope. 
+
+Forkchoice changes include:
+
+#### Constants
+
+| Name                 | Value       |
+| -------------------- | ----------- |
+| `PAYLOAD_TIMELY_THRESHOLD` | `PTC_SIZE/2` (=`uint64(256)`) | 
+| `INTERVALS_PER_SLOT` | `4` |
+| `PROPOSER_SCORE_BOOST` | `20` | 
+| `PAYLOAD_WITHHOLD_BOOST` | `40` | 
+| `PAYLOAD_REVEAL_BOOST` | `40` | 
+
+#### Containers
+
+```python
+class ChildNode(Container):
+    root: Root
+    slot: Slot
+    is_payload_present: bool
+```
+
+The class `LatestMessage` is modified to track the slot instead of an epoch:
+
+```python
+@dataclass(eq=True, frozen=True)
+class LatestMessage(object):
+    slot: Slot
+    root: Root
+```
+
+The class `Store` is modified to track the following fields
+```
+    execution_payload_states: Dict[Root, BeaconState] = field(default_factory=dict)
+    ptc_vote: Dict[Root, Vector[uint8, PTC_SIZE]] = field(default_factory=dict)
+    payload_withhold_boost_root: Root
+    payload_withhold_boost_full: bool
+    payload_reveal_boost_root: Root
+```
+
+- A new handler `on_payload_attestation_message` is called when receiving a `PayloadAttestationMessage` from the P2P network. 
+- A new handler `on_execution_payload` is called when receiving a `SignedExecutionPayloadEnvelope` from the P2P network. 
+
+### Engine API
+
+No changes needed
+
+## Rationale
+
+### Staked Builders
+
+Being a builder is a new attribution of validators. As such builders are staked in the beacon chain. This allows for in-protocol trustless enforcement of the builder's payment to the proposer. Alternatively, payment could be enforced in the Execution Layer at the cost of adding the corresponding EL consensus changing logic. Payments in the EL have the advantage of not requiring the builder to periodically submit deposit transactions to replenish their validator balance. Both systems require availability of funds before the payload is revealed: in the CL because the funds are staked, and in the EL because the payment much be done in the first transaction of the block. 
```

Comment:
"Being a builder is a new attribution of validators." Not sure I understand this sentence, I think you want to say "Validators are now able to perform the role of builders"?

Reply:
"Execution Layer (EL)"

Reply:
"consensus-changing"

Reply:
"in the Consensus Layer (CL)"

Reply:
"must be done"

Reply:
"Both systems require ..." this sentence is a bit confusing as written. The funds must be available because of the eventuality that a builder does not reveal their payload. *How* they are available depends on the layer (EL or CL).

Reply:
> "Being a builder is a new attribution of validators." Not sure I understand this sentence, I think you want to say "Validators are now able to perform the role of builders"?

Aren't these two sentences semantically equivalent?

Reply:
> "Both systems require ..." this sentence is a bit confusing as written. The funds must be available because of the eventuality that a builder does not reveal their payload. How they are available depends on the layer (EL or CL).

The point I am making is that the funds have to be available, it doesn't matter the layer. We do not know how to allow for payments with funds that are extracted from the block that the builder is constructed. This construct in principle can be done today with Mev Boost

Reply:
> > "Being a builder is a new attribution of validators." Not sure I understand this sentence, I think you want to say "Validators are now able to perform the role of builders"?
> 
> Aren't these two sentences semantically equivalent?

Yeah perhaps, I just wasn't 100% clear what the first meant, but apparently I did get it. It's more of a suggestion.

Reply:
> > "Both systems require ..." this sentence is a bit confusing as written. The funds must be available because of the eventuality that a builder does not reveal their payload. How they are available depends on the layer (EL or CL).
> 
> The point I am making is that the funds have to be available, it doesn't matter the layer. We do not know how to allow for payments with funds that are extracted from the block that the builder is constructed. This construct in principle can be done today with Mev Boost

It's just as written it sounds like "because the funds are staked, they must be available". I would just re-write a bit to make clear "The funds need to be available before the block is delivered/executed. In the CL, this is done by getting builders to stake. In the EL, this is done with a balance check and a payment transaction." (but how can this payment transaction go through if the payload is not delivered?)

Reply:
The issue is mostly that the transaction can be checked if it is TOB without executing the full payload. I'll add the rewording  you indicated here, plus a short comment regarding this

Reply:
But what if the builder never delivers the payload? Then there is no transaction no?

Reply:
The way that works in the EL is that the header contains the single tx that pays, it can be executed immediately, so that the CL transition has a simple execution transition by running that single tx.